actors/crypto/randomness.go

```diff
@@ -15,42 +16,55 @@ import (
 type DomainSeparationTag int
 
 const (
-	DomainSeparationTag_TicketDrawing DomainSeparationTag = 1 + iota
-	DomainSeparationTag_TicketProduction
+	DomainSeparationTag_TicketProduction DomainSeparationTag = 1 + iota
 	DomainSeparationTag_ElectionPoStChallengeSeed
 	DomainSeparationTag_SurprisePoStChallengeSeed
 	DomainSeparationTag_SurprisePoStSelectMiners
 	DomainSeparationTag_SurprisePoStSampleSectors
 )
 
 // Derive a random byte string from a domain separation tag and the appropriate values
-func DeriveRandWithMinerAddr(tag DomainSeparationTag, tix abi.RandomnessSeed, minerAddr addr.Address) abi.Randomness {
+func DeriveRandWithMinerAddr(tag DomainSeparationTag, randSeed []byte, minerAddr addr.Address) []byte {
 	var addrBuf bytes.Buffer
 	err := minerAddr.MarshalCBOR(&addrBuf)
 	autil.AssertNoError(err)
 
-	return _deriveRandInternal(tag, tix, -1, addrBuf.Bytes())
+	return _deriveRandInternal(tag, randSeed, addrBuf.Bytes())
 }
 
-func DeriveRandWithEpoch(tag DomainSeparationTag, tix abi.RandomnessSeed, epoch int64) abi.Randomness {
-	return _deriveRandInternal(tag, tix, -1, BigEndianBytesFromInt(epoch))
+type AddressEpochEntropy struct {
+	MinerAddress addr.Address // Must be an ID-addr
+	Epoch        abi.ChainEpoch
 }
 
-func _deriveRandInternal(tag DomainSeparationTag, randSeed abi.RandomnessSeed, index int, s []byte) abi.Randomness {
+func DeriveRandWithMinerAddrAndEpoch(tag DomainSeparationTag, randSeed []byte, minerAddr addr.Address, epoch abi.ChainEpoch) []byte {
+	entropy := &AddressEpochEntropy{
+		MinerAddress: minerAddr,
+		Epoch:        epoch,
+	}
+	var entrBuf bytes.Buffer
+	err := entropy.MarshalCBOR(&entrBuf)
+	autil.AssertNoError(err)
+
+	return _deriveRandInternal(tag, randSeed, entrBuf.Bytes())
+}
+
+func _deriveRandInternal(tag DomainSeparationTag, randSeed []byte, serializedEntropy []byte) []byte {
 	buffer := []byte{}
 	buffer = append(buffer, BigEndianBytesFromInt(int64(tag))...)
-	buffer = append(buffer, BigEndianBytesFromInt(int64(index))...)
 	buffer = append(buffer, randSeed...)
-	buffer = append(buffer, s...)
-	return SHA256(buffer)
+	buffer = append(buffer, serializedEntropy...)
+	bufHash := blake2b.Sum256(buffer)
```

Comment:
Blocking: the whole point of the hash method on the runtime is to avoid importing a specific hashing implementation here. You need to plumb the runtime's hash method through to here.

Reply:
It's a point well taken. Makes me think @ZenGround0's approach above of just abstracting this to sysCall and having this be node code makes sense, given these functions will also need to be called outside the VM (as part of generating a new block, ticket etc)... It needs to run the same way inside or outside the VM...

What do you think?